### DISCUSSION TITLE
How to define/customize Legend for a chart?

### DISCUSSION THREAD

**Discussion by kunjiangkun** [plotly/Plotly.NET#460] [General]
In version 5.0, legend is no longer an option for Layout.init. The legend option is also not exposed for trace such as Chart2D.Scatter. Looking at the code, it's defined for inner function Trace2DStyle.Scatter but Legend arg is set to null. Please provide an example for customizing Legend.

**Comment by kunjiangkun** [comment]
Somehow chart.WithLegend only supports arg of bool. Have to call GenericChartExtensions.WithLegend(trace, legendSetting) explicitly to get it working.

var boxPlot = box
    .WithLegend(true)
    .WithXAxis(xAxis)
    .WithLayout(plotLayout)
    .WithSize(plotWidth, plotHeight);
GenericChartExtensions.WithLegend(boxPlot, legendSetting);

**Comment by kMutagene** [comment]
It is hard for me to make out the problem because you are mixing some concepts here. Chart2D.Scatter does not create a trace - it creates a `GenericChart` which contains trace(s), a layout, a config, and display options. The reason is that plotly strictly separates the layout of a chart from traces. Traces are just the data and style options of a single trace (e.g., the marker style). Layout options on the other hand are styling options for the whole chart, no matter how many traces it has. A legend is exactly such an object - something that is displayed for one or more traces of the chart. 

So you are always aiming to set the legend on the layout, or use a method that changes the layout part of a `GenericChart`, for which you have multiple options:

- F# core API (these return F# functions, which you'd need to call `Invoke` on from C#):
  - `Layout.setLegend` or `Layout.updateLegendById` - static methods on the `Layout` class: https://github.com/plotly/Plotly.NET/blob/98a4d186642a2c67eba2779aa143218702ebbd66/src/Plotly.NET/Layout/Layout.fs#L1126
  - `Chart.withLegend` or `Chart.withLegendStyle` - static methods of the `Chart` API: https://github.com/plotly/Plotly.NET/blob/98a4d186642a2c67eba2779aa143218702ebbd66/src/Plotly.NET/ChartAPI/Chart.fs#L2629
  - `GenericChartExtensions` provides multiple overloads, one of them only takes boolean and needs to be removed, but there are others as well, see here : https://github.com/plotly/Plotly.NET/blob/98a4d186642a2c67eba2779aa143218702ebbd66/src/Plotly.NET/CSharpLayer/GenericChartExtensions.fs#L610-L613
 
- `GenericChartExtensions` in `Plotly.NET.C#`:
  - `WithLegend` or `WithLegendStyle`: https://github.com/plotly/Plotly.NET/blob/98a4d186642a2c67eba2779aa143218702ebbd66/src/Plotly.NET.CSharp/GenericChartExtensions.cs#L377
